x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ExpiredTokenRemover.java

```diff
@@ -43,15 +43,24 @@
     private final Client client;
     private final AtomicBoolean inProgress = new AtomicBoolean(false);
     private final TimeValue timeout;
+    private final SecurityIndexManager securityIndex;
+    private final SecurityIndexManager securityTokensIndex;
 
-    ExpiredTokenRemover(Settings settings, Client client) {
+    ExpiredTokenRemover(Settings settings, Client client, SecurityIndexManager securityIndex, SecurityIndexManager securityTokensIndex) {
         this.client = client;
         this.timeout = TokenService.DELETE_TIMEOUT.get(settings);
+        this.securityIndex = securityIndex;
+        this.securityTokensIndex = securityTokensIndex;
     }
 
     @Override
     public void doRun() {
-        DeleteByQueryRequest expiredDbq = new DeleteByQueryRequest(SecurityIndexManager.SECURITY_INDEX_NAME);
+        final DeleteByQueryRequest expiredDbq;
+        if (securityTokensIndex.exists()) {
+            expiredDbq = new DeleteByQueryRequest(securityTokensIndex.aliasName());
```

Comment:
is this truly a binary choice? Maybe the alias should point to both indices until the upgrade or something like that?

Reply:
It is a binary choice.

Tokens, expired or not, can sit in a single index at a certain moment in time: either `.security-6` or `.security-tokens-7`. When the remover job starts, it builds the query accordingly. It is possible that after building the query and before its execution, the tokens can move (from `.security-6` to `.security-tokens-7`) and the delete will miss them, but they will be catched by the next removal job.